packages/@lwc/template-compiler/src/codegen/scope.ts

```diff
@@ -9,35 +9,165 @@ import { walk } from 'estree-walker';
 import * as t from '../shared/estree';
 import { TEMPLATE_PARAMS } from '../shared/constants';
 import { isComponentProp } from '../shared/ir';
-import { IRNode, TemplateExpression } from '../shared/types';
+import { IRNode, TemplateExpression, TemplateIdentifier } from '../shared/types';
+import { NodeRefProxy } from './NodeRefProxy';
 
-/**
- * Bind the passed expression to the component instance. It applies the following transformation to the expression:
- * - {value} --> {$cmp.value}
- * - {value[index]} --> {$cmp.value[$cmp.index]}
- */
-export function bindExpression(expression: TemplateExpression, irNode: IRNode): t.Expression {
-    if (t.isIdentifier(expression)) {
-        if (isComponentProp(expression, irNode)) {
-            return t.memberExpression(t.identifier(TEMPLATE_PARAMS.INSTANCE), expression);
-        } else {
-            return expression;
+interface ComponentPropsUsageData {
+    name: string;
+    localName: string;
+    usage: NodeRefProxy;
+    instances: number;
+}
+
+function dumpScope(scope: Scope, body: t.Statement[], scopeVars: Map<string, string>) {
+    const nextScopeVars = new Map(scopeVars);
+    const variablesInThisScope: [string, string][] = [];
+
+    for (const [name, propData] of scope.usedProps) {
+        // the variable is already defined in the scope.
+        const generatedNameInScope = scopeVars.get(name);
+        if (generatedNameInScope !== undefined) {
+            propData.usage.swap(t.identifier(generatedNameInScope));
+        } else if (propData.instances > 1 || scope.aggregatedPropNamesUsedInChildScopes.has(name)) {
+            // name is not defined in the ancestor scopes, and is:
+            // a) used multiple times in this scope.
+            // b) used one time in this scope and at least once in one of the child scopes. Therefore
+            //    we can compute it in this scope.
+            const { localName } = propData;
+
+            nextScopeVars.set(name, localName);
+            propData.usage.swap(t.identifier(localName));
+            variablesInThisScope.push([name, localName]);
+        }
+    }
+
+    for (const childScope of scope.childScopes) {
+        body.unshift(childScope.scopeFn!);
```

Comment:
An assertion should be added here to make sure the `setFn` should be invoked before `serializeInto`